### PR TITLE
add roi_max_align_2d

### PR DESCRIPTION
**Pull request by knorth55**
~~Merge after #5070~~
~~[WIP] I haven't done GPU test yet.~~
Update: GPU tested.

add `roi_max_align_2d`.
#5070 is `average` pooling of `roi_align`, and this PR add `max` pooling of `roi_align`.
mainly `max` is for detection, and `average` for instance segmentation.

BTW, `roi_pooling_2d` is `max` pooling, but default `roi_align_2d` should be `average`?

cc. @wkentaro

### CODE REVIEW

**Comment by wkentaro** [comment]
What about switching the operation (`max` or `average`) by keyword argument to the `roi_align_2d` function?
For example, `roi_align_2d(x, rois, 7, 7, spatial_scale=0.125, operation='max')`.

And I agree that we set the `operation` to `'max'` by default.

**Comment by knorth55** [comment]
`max_pooling` and `average_pooling` are different named function, i think it is useful and easy to use add `roi_max_align` and `roi_average_align`.
How about adding both `roi_max_align` and `roi_average_align` function and add `roi_align(operation='max' or'average'` as interface?

**Comment by wkentaro** [comment]
That is fine for me. It is like `dialted_convolution_2d` (https://github.com/chainer/chainer/blob/da5bef64ef9e0d5d0b0c520530f5f3692c063bac/chainer/functions/connection/dilated_convolution_2d.py) which uses `convolution_2d(dilate=dilate)`.

**Comment by knorth55** [comment]
I updated this PR cince #5070 is merged.

- add `roi_max_align_2d`
- rename `roi_align_2d` added in #5070 to `roi_average_align_2d`
- add `roi_align_2d` with `operation` key
  - `operation` default is `max` because `roi_pooling_2d` is `max`.

**Comment by kmaehashi** [comment]
@mitmul Can you be a reviewer?

**Comment by knorth55** [comment]
i rebased and fix conflicts.

**Comment by knorth55** [comment]
We talked with Chainer team, and we discussed to modify function interface.

**Comment by knorth55** [comment]
I updated inteface as below.
roi_max_align_2d

**Comment by knorth55** [comment]
@Hakuyume can you review this PR?

File: chainer/functions/pooling/roi_max_align_2d.py [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+
+        return bottom_diff, None, None
+
+
+def roi_max_align_2d(
+        x, rois, roi_indices, outsize, spatial_scale, sampling_ratio=None
+):
+    """Spatial Region of Interest (ROI) max align function.
+
+    This function acts similarly to :class:`~functions.ROIPooling2D`, but
```
Comment (Hakuyume) [inline chainer/functions/pooling/roi_max_align_2d.py]: `functions.ROIPooling2D` was removed and `functions.roi_pooling_2d` is deprecated. How about pointing `functions.roi_max_pooling_2d` or `functions.roi_average_align_2d`?

Comment (knorth55) [reply chainer/functions/pooling/roi_max_align_2d.py]: I want to refer `functions.roi_max_poolind_2d`, but it is not merged yet :(

Comment (Hakuyume) [reply chainer/functions/pooling/roi_max_align_2d.py]: #5304 is merged now :tada: :tada: :tada: 
Could you merge master into this PR and update the docs?

Comment (knorth55) [reply chainer/functions/pooling/roi_max_align_2d.py]: Updated :innocent:

File: chainer/functions/pooling/roi_max_align_2d.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+from chainer import function
+from chainer.functions.pooling.roi_average_align_2d \
+    import _GET_BILINEAR_INTERP_KERNEL
+from chainer.functions.pooling.roi_average_align_2d \
+    import _get_bilinear_interp_params
+from chainer.utils import type_check
+
+
+def _pair(x):
+    if isinstance(x, collections.Iterable):
```
Comment (Hakuyume) [inline chainer/functions/pooling/roi_max_align_2d.py]: https://github.com/chainer/chainer/pull/5304#pullrequestreview-168224992

File: chainer/functions/pooling/roi_max_align_2d.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+            raise TypeError(
+                'outsize[0] must be positive integer: {}, {}'
+                .format(type(outh), outh))
+        if not (isinstance(outw, int) and outw > 0):
+            raise TypeError(
+                'outsize[1] must be positive integer: {}, {}'
+                .format(type(outw), outw))
+        if isinstance(spatial_scale, int):
+            spatial_scale = float(spatial_scale)
+        elif not (isinstance(spatial_scale, float) and spatial_scale > 0):
```
Comment (Hakuyume) [inline chainer/functions/pooling/roi_max_align_2d.py]: https://github.com/chainer/chainer/pull/5304#discussion_r229417576

File: tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_align_2d.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+from chainer import cuda
+from chainer import functions
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+from chainer.testing import condition
+
+
+def _pair(x):
+    if isinstance(x, collections.Iterable):
```
Comment (Hakuyume) [inline tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_align_2d.py]: https://github.com/chainer/chainer/pull/5304#discussion_r230000381

File: chainer/functions/pooling/roi_max_align_2d.py (hunk 3 above)
Comment (knorth55) [reply chainer/functions/pooling/roi_max_align_2d.py]: solved

**Comment by Hakuyume** [comment]
jenkins, test this please

**Pull request merged**